File: static/api/stats.php
<?php
// Stats API für DEL-Doku
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

// Funktion um Challenge-Datum aus Stats zu extrahieren
function getChallengeDate($stats) {
    if (isset($stats['currentChallenge'])) {
        return $stats['currentChallenge'];
    }
    return date('Y-m-d'); // Fallback: heutiges Datum
}

// Funktion um Stats-Datei für ein Challenge-Datum zu bekommen
function getStatsFile($challengeDate = null) {
    if (!$challengeDate) {
        $challengeDate = date('Y-m-d');
    }
    return __DIR__ . "/../../data/stats_$challengeDate.json";
}

// Stelle sicher, dass das data-Verzeichnis existiert
$dataDir = __DIR__ . '/../../data';
if (!file_exists($dataDir)) {
    mkdir($dataDir, 0755, true);
}

// OPTIONS Request für CORS
if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

// GET: Lade Stats
if ($_SERVER['REQUEST_METHOD'] === 'GET') {
    $userId = $_GET['userId'] ?? null;
    $challengeDate = $_GET['challengeDate'] ?? date('Y-m-d');
    
    $statsFile = getStatsFile($challengeDate);
    
    // Lade alle Stats für dieses Challenge-Datum
    $allStats = [];
    if (file_exists($statsFile)) {
        $content = file_get_contents($statsFile);
        $allStats = json_decode($content, true) ?: [];
    }
    
    // Wenn userId angegeben, gebe nur diesen User zurück
    if ($userId) {
        $userStats = $allStats[$userId] ?? null;
        echo json_encode([
            'userId' => $userId,
            'stats' => $userStats
        ]);
    } else {
        // Ohne userId: gebe alle Stats zurück (für Score-Berechnung)
        echo json_encode($allStats);
    }
    exit;
}

// POST: Speichere Stats
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $input = file_get_contents('php://input');
    $data = json_decode($input, true);
    
    if (!isset($data['userId']) || !isset($data['stats'])) {
        http_response_code(400);
        echo json_encode(['error' => 'userId and stats required']);
        exit;
    }
    
    $userId = $data['userId'];
    $stats = $data['stats'];
    
    // Extrahiere Challenge-Datum aus Stats
    $challengeDate = getChallengeDate($stats);
    $statsFile = getStatsFile($challengeDate);
    
    // Lade existierende Stats für dieses Challenge-Datum
    $allStats = [];
    if (file_exists($statsFile)) {
        $content = file_get_contents($statsFile);
        $allStats = json_decode($content, true) ?: [];
    }
    
    // Update Stats für diesen User
    $allStats[$userId] = $stats;
    
    // Speichere zurück
    file_put_contents($statsFile, json_encode($allStats, JSON_PRETTY_PRINT));
    
    echo json_encode([
        'success' => true,
        'userId' => $userId,
        'stats' => $stats,
        'challengeDate' => $challengeDate
    ]);
    exit;
}

http_response_code(405);
echo json_encode(['error' => 'Method not allowed']);

File: static/api/stats.py
#!/usr/bin/env python3
"""
Stats API für DELDoku (Python CGI)
"""
import cgi
import cgitb
import json
import os
import sys
from datetime import date
from pathlib import Path

# Aktiviere CGI-Fehlerausgabe für Debugging
cgitb.enable()

# Basis-Pfad für Stats-Dateien
DATA_DIR = Path(__file__).parent.parent.parent / 'data'


def get_challenge_date(stats):
    """Extrahiere Challenge-Datum aus Stats"""
    if isinstance(stats, dict) and 'currentChallenge' in stats:
        return stats['currentChallenge']
    return date.today().isoformat()


def get_stats_file(challenge_date=None):
    """Pfad zur Stats-Datei für ein Challenge-Datum"""
    if not challenge_date:
        challenge_date = date.today().isoformat()
    return DATA_DIR / f'stats_{challenge_date}.json'


def send_response(data, status=200):
    """Sende JSON-Antwort"""
    print("Content-Type: application/json")
    print("Access-Control-Allow-Origin: *")
    print("Access-Control-Allow-Methods: GET, POST, OPTIONS")
    print("Access-Control-Allow-Headers: Content-Type")
    print(f"Status: {status}")
    print()  # Leere Zeile zwischen Headers und Body
    print(json.dumps(data, indent=2))
    sys.exit(0)


def load_stats(challenge_date=None):
    """Lade Stats für ein Challenge-Datum"""
    stats_file = get_stats_file(challenge_date)
    
    if stats_file.exists():
        with open(stats_file, 'r', encoding='utf-8') as f:
            return json.load(f)
    return {}


def save_stats(stats, challenge_date=None):
    """Speichere Stats für ein Challenge-Datum"""
    stats_file = get_stats_file(challenge_date)
    
    # Erstelle Verzeichnis falls nicht vorhanden
    stats_file.parent.mkdir(parents=True, exist_ok=True)

    with open(stats_file, 'w', encoding='utf-8') as f:
        json.dump(stats, f, indent=2, ensure_ascii=False)


def handle_get():
    """GET Request: Lade Stats"""
    form = cgi.FieldStorage()
    user_id = form.getvalue('userId')
    challenge_date = form.getvalue('challengeDate')  # Optional: Challenge-Datum

    all_stats = load_stats(challenge_date)

    if user_id:
        # Gebe nur Stats für einen User zurück
        user_stats = all_stats.get(user_id)
        send_response({
            'userId': user_id,
            'stats': user_stats
        })
    else:
        # Gebe alle Stats zurück (für Score-Berechnung)
        send_response(all_stats)


def handle_post():
    """POST Request: Speichere Stats"""
    try:
        # Lese Request Body
        content_length = int(os.environ.get('CONTENT_LENGTH', 0))
        post_data = sys.stdin.read(content_length)
        data = json.loads(post_data)

        user_id = data.get('userId')
        stats = data.get('stats')

        if not user_id or not stats:
            send_response({'error': 'userId and stats required'}, 400)

        # Bestimme Challenge-Datum aus Stats
        challenge_date = get_challenge_date(stats)

        # Lade existierende Stats für dieses Challenge-Datum
        all_stats = load_stats(challenge_date)

        # Update Stats für diesen User
        all_stats[user_id] = stats

        # Speichere zurück in die challenge-spezifische Datei
        save_stats(all_stats, challenge_date)

        send_response({
            'success': True,
            'userId': user_id,
            'stats': stats,
            'challengeDate': challenge_date
        })
    except json.JSONDecodeError:
        send_response({'error': 'Invalid JSON'}, 400)
    except Exception as e:
        send_response({'error': str(e)}, 500)


def main():
    """Hauptfunktion"""
    method = os.environ.get('REQUEST_METHOD', 'GET')

    if method == 'OPTIONS':
        # CORS Preflight
        send_response({}, 200)
    elif method == 'GET':
        handle_get()
    elif method == 'POST':
        handle_post()
    else:
        send_response({'error': 'Method not allowed'}, 405)


if __name__ == '__main__':
    main()
